### PR TITLE
Add support for two-way websocket routes. Fixes #6130

### LINKED ISSUE #6130: Add support for websockets two-way routes
This feature was already added to serverless-websocket-plugin, can you add it to the framework?

> Adds support for two-way routes by using the **routeResponseSelectionExpression** property. When this property is set, RouteResponseSelectionExpression will be set on the Route and a RouteResponse is also created.

### PR DESCRIPTION
**Pull request by DougHamil**
## What did you implement

This PR adds support for two-way websocket connections for the API Gateway. It adds the `routeResponseSelectionExpression` parameter to the `websocket` event, which will set the `RouteResponseSelectionExpression` property on the Route and create a `RouteResponse`

Closes #6130 

## How can we verify it

Here's an example configuration:

```yml
functions:
  hello:
    handler: handler.helloHandler
    events:
      - websocket:
          route: hello
          routeResponseSelectionExpression: $default

```

## Todos

<details>
<summary>Useful Scripts</summary>

- `npm run test:ci` --> Run all validation checks on proposed changes
- `npm run lint:updated` --> Lint all the updated files
- `npm run lint:fix` --> Automatically fix lint problems (if possible)
- `npm run prettier-check:updated` --> Check if updated files adhere to Prettier config
- `npm run prettify:updated` --> Prettify all the updated files

</details>

- [x] Write and run all tests
- [ ] Write documentation
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

### CODE REVIEW

File: lib/plugins/aws/package/compile/events/websockets/lib/routeResponses.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+      const websocketsRouteResponseLogicalId = this.provider.naming.getWebsocketsRouteResponseLogicalId(
+        event.route
+      );
+
+      const websocketsRouteLogicalId = this.provider.naming.getWebsocketsRouteLogicalId(
+        event.route
+      );
+
+      _.merge(this.serverless.service.provider.compiledCloudFormationTemplate.Resources, {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/websockets/lib/routeResponses.js]: Let's use `Object.assign` (which tend to resort to lodash, only when there's no convenient native counterpart, and here deep assign is not justified)

File: lib/plugins/aws/package/compile/events/websockets/lib/routeResponses.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            RouteId: {
+              Ref: websocketsRouteLogicalId,
+            },
+            RouteResponseKey: '$default',
+          },
+        },
+      });
+    });
+
+    return BbPromise.resolve();
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/websockets/lib/routeResponses.js]: Let's remove this. It's a sync function, no need to return promise.

File: lib/plugins/aws/package/compile/events/websockets/lib/validate.js [hunk 3]
```diff
@@ -31,6 +31,12 @@ module.exports = {
               route: event.websocket.route,
             };
 
+            // route response
+            if (_.isString(event.websocket.routeResponseSelectionExpression)) {
```
Comment (medikoo) [inline lib/plugins/aws/package/compile/events/websockets/lib/validate.js]: Let's do `if (event.websocket.routeResponseSelectionExpression)`

**Comment by DougHamil** [comment]
@medikoo I updated the websockets integration-test to include a two-way route test, which works when I run it locally. Please let me know if there's anything else I should look into.

File: tests/integration-all/websocket/tests.js [hunk 4]
```diff
@@ -43,12 +43,54 @@ describe('AWS - API Gateway Websocket Integration Test', function() {
...
+        ws.on('open', () => {
+          console.info("Sending message to 'hello' route");
+          ws.send(JSON.stringify({ action: 'hello', name: 'serverless' }));
+        });
+
+        ws.on('close', resolve);
+
+        ws.on('message', event => {
+          try {
+            console.info('Received WebSocket message', event);
```
Comment (medikoo) [inline tests/integration-all/websocket/tests.js]: Let's use `log` library as here: https://github.com/serverless/serverless/blob/1793cf8d7a55b85fc6505ae493dcca2292e443d2/tests/integration-all/api-gateway/tests.js#L188

Then log is output only when with adequate log level setting, or when tests fail, which makes it more DX friendly

Here configured log, could look as:

```javascript
log.debug('Received WebSocket message %o', event)
```

**Comment by waynerobinson** [comment]
Hi everyone.

What’s left to do to get this merged?

Cheers

**Comment by medikoo** [comment]
@DougHamil is this ready for take? (if so, please re-request review, thank you!)

**Comment by DougHamil** [comment]
Sure thing! I've added some documentation, please let me know if you have any edits or suggestions on it.

**Pull request merged**